cli/crud/lua/import.lua

```diff
@@ -0,0 +1,467 @@
+        --         tupels_for_import
+        -- )
+    end
+end
+
+
+-- Due to the current design of crud batching, the equal tuples are indistinguishable
+-- in the variables RES/ERR and uploaded tuples at router.
+-- It is a temporary solution with this function,
+-- that using a matching search in RES/ERR variables and uploaded tuples.
+-- As the matches are found, the values from the RES/ERR are crossed out.
+-- As a result, we won't lose any tuple, but error messages, for exaple, can be mistakenly swapped between tuples!
+local init_eval_is_tuples_equal = function()
+    box.session.storage.crud_import_is_tuples_equal = function (tuple1, tuple2)
+        if tuple1 == nil or tuple2 == nil then
+            return false
+        end
+        if #tuple1 ~= #tuple2 then
+            return false
+        end
+        for tuple_index, _ in pairs(tuple1) do
+            if tuple1[tuple_index] ~= tuple2[tuple_index] then
+                return false
+            end
+        end
+        return true
+    end
+end
+
+-- Allows to get the batch with updated context after import.
+local init_eval_get_batch_final_ctx = function()
+    box.session.storage.crud_import_get_batch_final_ctx = function ()
+        -- Case of empty tables.
+        if box.session.storage.batch_insert_res == nil then
+            box.session.storage.batch_insert_res = {}
+            box.session.storage.batch_insert_res['rows'] = {}
+        end
+        if box.session.storage.batch_insert_err == nil then
+            box.session.storage.batch_insert_err = {}
+        end
+
+        local batch_final_ctx = box.session.storage.crudimport_batch_table_for_import
+        for parsed_tuple_num, _ in pairs(box.session.storage.crudimport_batch_table_for_import['tuples']) do
+            for imported_tuple_num, imported_tuple_val in pairs(box.session.storage.batch_insert_res['rows']) do
+                if box.session.storage.crud_import_is_tuples_equal(
+                    imported_tuple_val,
+                    box.session.storage.crudimport_batch_table_for_import
+                        ['tuples'][parsed_tuple_num]['crudCtx']['castedTuple']
+                    ) and
+                    box.session.storage.crudimport_batch_table_for_import
+                        ['tuples'][parsed_tuple_num]['crudCtx']['imported'] == false
+                    then
+                        batch_final_ctx['tuples'][parsed_tuple_num]['crudCtx']['imported'] = true
+                        -- Try to resist ambiguity, but a swap is still possible (at least we don't lose record).
+                        box.session.storage.batch_insert_res['rows'][imported_tuple_num] = nil
+                        goto resulted_tuple_found
+                end
+            end
+            for error_tuple_num, error_tuple_val in pairs(box.session.storage.batch_insert_err) do
+                if box.session.storage.crud_import_is_tuples_equal(
+                    error_tuple_val['tuple'],
```

Comment:
It was renamed to `operation_data` in the recent `crud.*_many()` implementation, because it may be an object for `crud.*_object_many()`. Please, update accordingly.

Reply:
Done